_posts/2019-11-14-diagnosing-our-radiation-problems.markdown

```diff
@@ -0,0 +1,118 @@
+---
+
+Email! Electronic mail! What a concept! Like many companies, Artsy has built products on top of email, but this is
+a decision that (like many companies) Artsy periodically regrets. But overall, our email systems work well!
+
+But what about when it doesn't? Well that's what today's blog post is about: what happens when things break and you
+don't know why?
+
+<!-- more -->
+
+I have learned a lot
+[since my first on-call shift](https://artsy.github.io/blog/2018/05/30/my-first-week-on-call/), but going on call
+still gives me a little stage fright. As I start a shift, I'm on the look-out for things that might break, and soon
+after starting a recent on-call shift, "breaking" is exactly what things did.
+
+![Screenshot of Slack alert](/images/2019-11-07-diagnosing-our-radiation-problems/alert.png)
+
+We got an automated alert on Slack that one of our email services, code-named "Radiation", had really high response
+times. More specifically, it was timing out. Yikes. The next twenty two hours was a deep dive into Heroku, New
+Relic, Rails, and Postgres, all to isolate the problem and produce a solution.
+
+The nice thing about email, as a protocol, is that it's based on a store-and-forward concept. That means that if a
+message delivery fails, email servers will try again later (typically with an exponential backoff). While the
+server was unresponsive, email deliveries would be delayed, but the emails themselves were not dropped. We wasted
+no time addressing the problem, but we were also confident that once we fixed the issue, the data would be okay.
+
+After the alert, Ashkan (Radiation's point-person) and I communicated with our colleagues (engineers and affected
+non-engineers) about the problem. With the help of Chung-Yi, we began investigating the immediate problem (with a
+focus on _mitigating_ it, rather than necessarily _fixing_ it). Oh, we tried it all: spinning up more Heroku Dynos
+to handle requests, increasing concurrency on the individual Dynos, restarting Redis and Postgres stores. Each
+time, things would work briefly before the Radiation API would begin timing out again. Hmm. We started examining
+the Radiation code and database (keeping our _ion the prize_: mitigation).
+
+Ashkan investigated slow database queries. We added new Postgres indexes to speed up queries, and restructured
+others to avoid expensive joins. Unfortunately, all this accomplished was extending the time after a Dyno reboot
+that things would work (before beginning to timeout again). Because
+[Artsy stores emails as encrypted-at-rest](https://artsy.github.io/blog/2017/05/30/database-encryption/), it was
+difficult to pinpoint which exact message could be causing the timeouts. And Radiation itself didn't have much in
+the way of logging, which would normally help us debug the problem.
+
+It was frustrating, to try to fix something, but only manage to improve how long it took to break again. We had
```

Comment:
I think this might read a bit better without the two commas.

Reply:
My high school English teacher was _always_ after me about my comma-splices 😅